extensions/vault/runtime/src/main/java/io/quarkus/vault/VaultProxyImpl.java

```diff
@@ -0,0 +1,150 @@
+
+public class VaultProxyImpl implements VaultBaseProxy {
+
+    private static final Logger log = Logger.getLogger(VaultProxyImpl.class);
+
+    private CloseableHttpClient httpclient;
+    private Jsonb jsonb = JsonbBuilder.create();
+    private String url;
+
+    public VaultProxyImpl(CloseableHttpClient httpclient, String url) {
+        this.httpclient = httpclient;
+        this.url = url;
+    }
+
+    @Override
+    public VaultInit init(VaultInitBody body) {
+        return put("/v1/sys/init", body, VaultInit.class);
+    }
+
+    @Override
+    public VaultUserPassAuth loginUserPass(String user, VaultUserPassAuthBody body) {
+        return post("/v1/auth/userpass/login/" + user, body, VaultUserPassAuth.class);
+    }
+
+    @Override
+    public VaultKubernetesAuth loginK8s(VaultKubernetesAuthBody body) {
+        return post("/v1/auth/kubernetes/login", body, VaultKubernetesAuth.class);
+    }
+
+    @Override
+    public VaultKvSecretV1 getSecretV1(String token, String secretEnginePath, String path) {
+        return get("/v1/" + secretEnginePath + "/" + path, token, VaultKvSecretV1.class);
+    }
+
+    @Override
+    public VaultKvSecretV2 getSecretV2(String token, String secretEnginePath, String path) {
+        return get("/v1/" + secretEnginePath + "/data/" + path, token, VaultKvSecretV2.class);
+    }
+
+    @Override
+    public VaultHealth getHealth() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public VaultSealStatus getSealStatus() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public VaultRenewSelf renewSelf(String token, VaultRenewSelfBody body) {
+        return post("auth/token/renew-self", body, VaultRenewSelf.class);
+    }
+
+    @Override
+    public VaultLookupSelf lookupSelf(String token) {
+        return get("auth/token/lookup-self", token, VaultLookupSelf.class);
+    }
+
+    @Override
+    public VaultLeasesLookup lookupLease(String token, VaultLeasesBody body) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public VaultRenewLease renewLease(String token, VaultLeasesBody body) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public VaultCreds generateCredentials(String token, String dbrole) {
+        throw new UnsupportedOperationException();
+    }
+
+    // ---
+
+    private <T> T get(String path, String token, Class<T> resultClass) {
+        return exec(new HttpGet(url + path), token, null, resultClass);
+    }
+
+    private <T> T post(String path, Object body, Class<T> resultClass) {
+        return exec(new HttpPost(url + path), null, body, resultClass);
+    }
+
+    private <T> T put(String path, Object body, Class<T> resultClass) {
+        return exec(new HttpPut(url + path), null, body, resultClass);
+    }
+
+    private <T> T exec(HttpRequestBase target, String token, Object body, Class<T> resultClass) {
```

Comment:
May be I'm missing something. Why it is even needed this code ? MP REST client, via the `VaultProxy` should just work, and take care itself of all the serialization/etc ?

Reply:
there are 2 use cases:
 - use vault directly from application and/or from agroal => with use the MP rest client
 - use vault as a configsource => we use httpclient
because I could not get the VaultConfigSource to use MP rest client. configsources are executed very early, before the rest client is initialized. what happens if I try to use the rest client from the VaultConfigSource (during configsources initialization) is that the rest client tries to access some properties from MP config. 
it is my uderstanding that all config sources have to be executed, before and of the extensions that use MP config can be used (otherwise we have a chicken/egg problem).
if there was a way, that would be great, as I would prefer not having to reimplement operations.